pkgs/development/beam-modules/build-mix.nix

```diff
@@ -0,0 +1,89 @@
+{ stdenv, writeText, elixir, erlang, hex, lib }:
+
+{ name
+, version
+, src
+, buildInputs ? [ ]
+, beamDeps ? [ ]
+, propagatedBuildInputs ? [ ]
+, postPatch ? ""
+, compilePorts ? false
+, meta ? { }
+, enableDebugInfo ? false
+, mixEnv ? "prod"
+, ...
+}@attrs:
+
+with lib;
+let
+  shell = drv: stdenv.mkDerivation {
+    name = "interactive-shell-${drv.name}";
+    buildInputs = [ drv ];
+  };
+
+  pkg = self: stdenv.mkDerivation (attrs // {
+    name = "${name}-${version}";
+    inherit version;
+    inherit src;
+
+    MIX_ENV = mixEnv;
+    MIX_DEBUG = if enableDebugInfo then 1 else 0;
+    HEX_OFFLINE = 1;
+
+    setupHook = attrs.setupHook or
+      writeText "setupHook.sh" ''
+      addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
```

Comment:
This is really just a nitpick, but dependencies might be added several times.
how about
```
# some dependencies can have themselves common dependencies
if ! [[ $ERL_LIBS == *"$1"* ]] ; then
  addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
fi
```

Reply:
I don't understand how packages would be added to the search path more than once. If multiple packages have same dependency, that dependency would only be installed once and the setupHook would only run once.

I just tested that by adding jason as a dependency for several packages, made each of those packages dependencies of a custom package. Built it. Looked inside `result/nix-support/propagated-build-inputs`. Jason was only listed once.

Reply:
Just to confirm, I ran `nix-shell` in the project directory and confirmed that jason was only listed once in $ERL_LIBS.

Reply:
I had added a debug line to see which dependencies where being imported
```
SETUP HOOK FOR /nix/store/37il3svijbcj67k2lfzz9hmwc9wdnqsl-ssl_verify_fun-1.1.6
SETUP HOOK FOR /nix/store/0w9i4f9rd39ap41six42rf6xdqv8ryaa-jason-1.2.2
SETUP HOOK FOR /nix/store/plcb89d9mkxnil0fgfaj5cncia7209ss-decimal-2.0.0
SETUP HOOK FOR /nix/store/37il3svijbcj67k2lfzz9hmwc9wdnqsl-ssl_verify_fun-1.1.6
SETUP HOOK FOR /nix/store/37il3svijbcj67k2lfzz9hmwc9wdnqsl-ssl_verify_fun-1.1.6
SETUP HOOK FOR /nix/store/0w9i4f9rd39ap41six42rf6xdqv8ryaa-jason-1.2.2
SETUP HOOK FOR /nix/store/plcb89d9mkxnil0fgfaj5cncia7209ss-decimal-2.0.0
SETUP HOOK FOR /nix/store/pq4zbzymfjssgfw961ihlwc185ibnn81-uri_query-0.1.2
SETUP HOOK FOR /nix/store/37il3svijbcj67k2lfzz9hmwc9wdnqsl-ssl_verify_fun-1.1.6
SETUP HOOK FOR /nix/store/0w9i4f9rd39ap41six42rf6xdqv8ryaa-jason-1.2.2
SETUP HOOK FOR /nix/store/plcb89d9mkxnil0fgfaj5cncia7209ss-decimal-2.0.0
SETUP HOOK FOR /nix/store/dm1jas9bacbd69l8k8x2xdwzv5yivzl1-plug-1.11.1
SETUP HOOK FOR /nix/store/pqhc5dp2l0rvcwgzdxx0qwkppg5whsj4-mime-1.6.0
SETUP HOOK FOR /nix/store/njlns3zbl9wr9kvmgsc8x7jvywz04hak-plug_crypto-1.2.2
SETUP HOOK FOR /nix/store/1kgn5nzp0g2q94f3jknd4h864fnl3cas-plug_cowboy-2.5.0
```
As I was seeing stuff appear twice in there i assumed the ERL_LIBS were being duplicated.
echoing the ERL_LIBS confirmed it wasn't. The logs are just stuff running on different cores. My bad.